### ISSUE TITLE
Backport fuzzy/BCE date implementation from Arches v.4 to v.3

### ISSUE THREAD

**Issue by azerbini** [azerbini/eamena_dev#31]
_Problem_

For archaeological time spans, we need to be able to implement a date object that allows us to encompass fuzzy as well as negative (i.e. BC) dates. This is to address:  
https://github.com/azerbini/eamena_v3/issues/39

_Solution_

A solution would be to backport Arches v.4 solution for this issue, discussed here: https://github.com/archesproject/arches/issues/2337

**Comment by mradamcox** [comment]
@azerbini I'm afraid that the v4 implementation can't be backported to v3 very easily, because it depends on some internal structures (DataTypes and Widgets) that aren't present in v3. I'm looking into using the fuzzy date implementation that @coherit put together, https://github.com/coherit/edtf-arches-hip.

**Comment by azerbini** [comment]
Thanks Adam. An essential requirement for us would be to support negative (BCE) dates. I wonder whether we shouldn’t use Astropy instead? http://docs.astropy.org/en/stable/index.html

**Comment by mradamcox** [comment]
Astropy looks promising, but unfortunately it only works with python 3.5 and 3.6 http://docs.astropy.org/en/stable/install.html.

Still working on this though. It's a multi-faceted challenge as the dates need to be parsed in js, python and properly stored in elasticsearch. I'm thinking ES may be the biggest obstacle...

**Comment by mradamcox** [comment]
Ok, I've done a ton of work on this, following off of the @coherit work, and what I have will work as follows:

+ we'll need to add new nodes to hold the extended dates (@azerbini said this would be acceptable)
    for these nodes, the business table name must be `string`
+ these nodes will have a free text input in the forms, and valid date values will be
    `YYYY-MM-DD` as before, `YYYY-MM`, or `YYYY`.
+ to enter a BCE date, just precede the date with `-`. For example, `-44-03-15` or `-12000`.
+ **all** dates are now indexed into two new ES mappings `extendeddates` and `extendeddategroups`, where they are stored as integers
+ the time-filter will work as before, now querying the new ES mappings instead of the old `dates` and `date_groups` mappings.

What I need now is to get the new nodes created locally so I can fully port and test my updates into the EAMENA code. @AzadehVafadari and @azerbini, any thoughts on that?

**Comment by azerbini** [comment]
Hi Adam,

Terrific work. What we will do is actually covert the Archaeological date nodes to string nodes, rather than creating new nodes.

**Comment by mradamcox** [comment]
Ok, I'm thinking these two nodes: `ARCHAEOLOGICAL_FROM_DATE.E61` and `ARCHAEOLOGICAL_TO_DATE.E61`

Will there be existing data stored in these nodes that will need to be migrated?

**Comment by azerbini** [comment]
No, those nodes don’t have any prior data. We are starting anew with those.

Sent from my iPhone

On 27 Jun 2018, at 18:53, mradamcox <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok, I'm thinking these two nodes: ARCHAEOLOGICAL_FROM_DATE.E61 and ARCHAEOLOGICAL_TO_DATE.E61

Will there be existing data stored in these nodes that will need to be migrated?

**Comment by mradamcox** [comment]
As I'm incorporating this work into EAMENA, it seems to me like the time filter is not added to search when multiple search terms are being entered. Is that correct?

I'm referring to the aggregated searches that you can create by enabling this button 
![image](https://user-images.githubusercontent.com/10427268/42004169-4e24f48c-7a34-11e8-982f-21027ab694bc.png)

**Comment by azerbini** [comment]
No that is not correct. The time filter can be activated by clicking on the clock next to the search box. Is it not working for you?

Sent from my iPhone

On 28 Jun 2018, at 01:58, mradamcox <notifications@github.com<mailto:notifications@github.com>> wrote:


As I'm incorporating this work into EAMENA, it seems to me like the time filter is not added to search when multiple search terms are being entered. Is that correct?

**Comment by azerbini** [comment]
The ‘aggregated’ searches are the advanced search toolkit. Each set of searching has a time filter component which, if triggered, activates the time filter within that search context

**Comment by mradamcox** [comment]
It works on the front-end but I wasn't seeing any new content in the time filter parameter of the request. However, I think I need to go back and look at how it was originally implemented before I made modifications to accommodate the new extended dates. Thanks for the response.

**Comment by mradamcox** [comment]
Alright, after looking back at the original configuration I found what I had lost and fixed things up to re-enabled the multiple time filters. It's still easy to get strange results by combining multiple start/end dates with the slider within a single time filter, but that is really more of a front-end and overall design consideration than it is related to this extended date effort.

**Comment by mradamcox** [comment]
Ah, I'm seeing now that https://github.com/azerbini/eamena_v3/issues/70 would actually enhance the date filter, and would be a good companion to this effort. Once this PR has been merged, it may be wise to put forth a single effort to handle issue 70 from the v3 repo.
